http/src/main/java/io/micronaut/http/body/RawMessageBodyHandlerRegistry.java

```diff
@@ -0,0 +1,352 @@
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.Experimental;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.io.buffer.ByteBuffer;
+import io.micronaut.core.io.buffer.ByteBufferFactory;
+import io.micronaut.core.io.buffer.ReferenceCounted;
+import io.micronaut.core.type.Argument;
+import io.micronaut.core.type.Headers;
+import io.micronaut.core.type.MutableHeaders;
+import io.micronaut.http.HttpHeaders;
+import io.micronaut.http.MediaType;
+import io.micronaut.http.codec.CodecException;
+import io.micronaut.runtime.ApplicationConfiguration;
+import jakarta.inject.Singleton;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Base class for {@link MessageBodyHandlerRegistry} that handles caching and exposes the raw
+ * handlers (String, byte[] and such).
+ *
+ * @author Graeme Rocher
+ * @since 4.0.0
+ */
+@Internal
+@Experimental
+abstract class RawMessageBodyHandlerRegistry implements MessageBodyHandlerRegistry {
+    private static final MessageBodyReader<Object> NO_READER = new NoReader();
+    private static final MessageBodyWriter<Object> NO_WRITER = new NoWriter();
+    private final Map<HandlerKey<?>, MessageBodyReader<?>> readers = new ConcurrentHashMap<>(10);
+    private final Map<HandlerKey<?>, MessageBodyWriter<?>> writers = new ConcurrentHashMap<>(10);
+
+    private final List<RawEntry> rawHandlers;
+
+    /**
+     * Default constructor.
+     *
+     * @param rawHandlers Handlers for raw values
+     */
+    RawMessageBodyHandlerRegistry(List<RawMessageBodyHandler<?>> rawHandlers) {
+        this.rawHandlers = rawHandlers.stream().flatMap(h -> h.getTypes().stream().map(t -> new RawEntry(t, h))).toList();
+    }
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    private <T> MessageBodyHandler<T> rawHandler(Argument<T> type, boolean covariant) {
+        for (RawEntry entry : rawHandlers) {
+            if (covariant ? entry.type.isAssignableFrom(type.getType()) : entry.type == type.getType()) {
+                return (MessageBodyHandler<T>) entry.handler;
+            }
+        }
+        return null;
+    }
+
+    protected abstract <T> MessageBodyReader<T> findReaderImpl(Argument<T> type, List<MediaType> mediaTypes);
+
+    @SuppressWarnings({"unchecked"})
+    @Override
+    public <T> Optional<MessageBodyReader<T>> findReader(Argument<T> type, List<MediaType> mediaTypes) {
+        HandlerKey<T> key = new HandlerKey<>(type, mediaTypes);
+        MessageBodyReader<?> messageBodyReader = readers.get(key);
+        if (messageBodyReader == null) {
```

Comment:
Would it make sense to have two maps:
- 1. All readers/writers contributed at the startup - simple hashmap - fast, the most common scenario
- 2. Precalculated readers/writers - requires a concurrent map

The minus is that we would need to access two maps in the worse scenario.
If the readers/writers are cached in the route info, this is unnecessary.

Reply:
the problem is that the matching is kind of fuzzy. for example the json mapper matches for any body type, and is cached separately for each type. so adding it for the base type (Object) is basically pointless.

The map is kind of like the cache in ConversionService.